packages/expo-modules-core/ios/ReactDelegates/EXReactRootViewFactory.mm

```diff
@@ -47,25 +56,47 @@ + (UIView *)createDefaultReactRootView:(nullable NSURL *)_bundleURL
   NSURL *bundleURL = _bundleURL ?: appDelegate.bundleURL;
   NSString *moduleName = _moduleName ?: appDelegate.moduleName;
   NSDictionary *initialProperties = _initialProperties ?: appDelegate.initialProps;
+
+   if (![appDelegate.rootViewFactory isKindOfClass:EXReactRootViewFactory.class]) {
+     @throw [NSException exceptionWithName:NSInternalInconsistencyException
+                                    reason:@"The appDelegate.rootViewFactory must be an EXReactRootViewFactory instance."
+                                  userInfo:nil];
+   }
+  EXReactRootViewFactory *appRootViewFactory = (EXReactRootViewFactory *)appDelegate.rootViewFactory;
+  RCTRootViewFactoryConfiguration *appRootViewFactoryConfiguration = [appRootViewFactory valueForKey:@"_configuration"];
+
   RCTRootViewFactoryConfiguration *configuration =
   [[RCTRootViewFactoryConfiguration alloc] initWithBundleURL:bundleURL
                                               newArchEnabled:appDelegate.fabricEnabled
                                           turboModuleEnabled:appDelegate.turboModuleEnabled
                                            bridgelessEnabled:appDelegate.bridgelessEnabled];
+  configuration.createRootViewWithBridge = appRootViewFactoryConfiguration.createRootViewWithBridge;
+  configuration.createBridgeWithDelegate = appRootViewFactoryConfiguration.createBridgeWithDelegate;
 
-  __weak RCTAppDelegate *weakDelegate = appDelegate;
-  configuration.createRootViewWithBridge = ^UIView *(RCTBridge *bridge, NSString *moduleName, NSDictionary *initProps)
-  {
-    return [weakDelegate createRootViewWithBridge:bridge moduleName:moduleName initProps:initProps];
-  };
+  EXReactRootViewFactory *factory = [[EXReactRootViewFactory alloc] initWithConfiguration:configuration andTurboModuleManagerDelegate:appDelegate];
+  UIView *rootView = [factory superViewWithModuleName:moduleName initialProperties:initialProperties launchOptions:launchOptions];
 
-  configuration.createBridgeWithDelegate = ^RCTBridge *(id<RCTBridgeDelegate> delegate, NSDictionary *launchOptions)
-  {
-    return [weakDelegate createBridgeWithDelegate:delegate launchOptions:launchOptions];
-  };
+  // We want to retain RCTHost/RCTBridge instance from the root view.
+  // RCTBridge is retained from RCTRootView by nature and RCTHost would be retained using associated objects.
+  // We will also release the unused RCTHost/RCTBridge from RCTRootViewFactory.
+  if (appDelegate.bridgelessEnabled) {
+    RCTHost *rctHost = [factory valueForKey:@"_reactHost"];
+    objc_setAssociatedObject(rootView, "_reactHost", rctHost, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+    [appRootViewFactory setValue:nil forKey:@"_reactHost"];
```

Comment:
🙉 
I wish we don't have to do this...

Reply:
let me think more about this. probably we need to change something for upstream.
it works for greenfield apps because typically it's 1-1 mapping for AppDelegate <-> RCTRootViewFactory. however, we have dev-menu and dev-launcher that would have different RCTHost instances other than the one from AppDelegate. maybe we should also have a RCTRootViewFactory inside their host.

Reply:
@tsapeta does that look better to you? https://github.com/expo/expo/pull/27715/commits/63ff5c9f4fecd04a16c07b500c0b8c626a30a037

Reply:
later we could remove the RCTRootViewFactory internal category when https://github.com/facebook/react-native/pull/43528 is available. i'll keep it as-it and remove the internal category later.